pallets/gear/src/lib.rs

```diff
@@ -260,14 +265,66 @@ pub mod pallet {
             .ok()
         }
 
+        fn decrease_gas_allowance(gas_charge: u64) {
+            GasAllowance::<T>::mutate(|x| {
+                *x = x.saturating_sub(gas_charge)
+            });
+        }
+
+        /// Returns true if a program has been successfully initialized
+        pub fn is_initialized(program_id: H256) -> bool {
+            common::get_program_state(program_id)
+                .map(|s| matches!(s, ProgramState::Initialized))
+                .unwrap_or(false)
+        }
+
         /// Returns true if a program resulted in an error during initialization
         /// but hasn't been explicitly removed from storage by its creator
-        pub fn is_uninitialized(program_id: H256) -> bool {
+        pub fn is_failed(program_id: H256) -> bool {
             ProgramsLimbo::<T>::get(program_id)
                 .map(|_| true)
                 .unwrap_or(false)
         }
 
+        /// Internal helper function. Moves messages from waiting list to the queue. Returns a flag if there are
+        /// more (probably) messages and used weight.
+        fn requeue_waiting_messages(
+            program_id: H256,
+            available_weight: Weight,
+        ) -> (Option<WaitingMessages>, Weight) {
+            let one_step_weight = T::DbWeight::get().reads(2) + T::DbWeight::get().writes(1);
+            let message_count = available_weight / one_step_weight;
+
+            let actual_count = common::WaitingMessageIterator::drain(program_id, None)
+                .take(message_count as usize)
+                .fold(0u64, |i, (m, _)| {
+                    common::queue_message(m);
+                    i + 1
+                });
+
+            let used_weight = actual_count * one_step_weight;
+            if actual_count < message_count {
+                (None, used_weight)
+            } else {
+                (Some(WaitingMessages), used_weight)
+            }
+        }
+
+        /// Tries to wake all messages sent to `destination_program_id` while it hasn't been initialized.
+        pub fn wake_waiting_messages(destination_program_id: H256) {
+            let (remained, used_weight) = Self::requeue_waiting_messages(
+                destination_program_id,
+                Self::gas_allowance() as Weight,
+            );
+            Self::decrease_gas_allowance(used_weight);
```

Comment:
@ekovalev 

I realized that we have two transformations: (1) gas -> fee, (2) gas -> weight. Here the identity mappging `gas -> weight` is assumed which is not correct in general

Reply:
We also have an explicit `WeightToFee` mapping from `Weigth` to `BalanceOf<T>` so we need to make sure this diagram always commutes. As long as all the maps are Identity we are fine, but generally I guess we should always express the third as a composition of the first two.

Reply:
Care to create a task for proper conversion traits implementation and mention in a `TODO` that `Self::gas_allowance() as Weight` will be replaced with a more appropriate gas to weight conversion function?